### ISSUE TITLE
Wrong dockers used in various cases

### ISSUE THREAD

**Issue by ckcollab** [codalab/codalab-competitions#2034]
When a submission is run in a docker, we should always use the same docker to run it:
- if we re-run it (regardless of whether the organizer changes the default docker)
- if the submission is migrated to another phase

If people what to change the docker of a submission, they should download it and re-submit is with a different docker.

**Comment by madclam** [comment]
Related to that: when you re-run a submission, it does not use the same docker. Therefore it is not reproducible.

**Comment by madclam** [comment]
It is not possible to know how/where the ingestion program is run in the metadata
![image](https://user-images.githubusercontent.com/7594063/32526404-f83f2b3a-c3dd-11e7-8fca-e42ef94bd69e.png)

In this example, submission 577 used a docker madclam/codalab

**Comment by madclam** [comment]
## CLARIFICATIONS ON DOCKER SPECS:

(1) Admin dockers
- run codalab front-end
- run computer workers
- supply a default docker to run organizers’ and participants’ code = ckcollab/codalab-legacy

(2) Organizers’ dockers
(I)- ingestion program
(P)- participant submissions
(S)- scoring program
the are supplied in the editor:
+ Ingestion program docker image:
+ Default docker image: (participant submissions???)
==> Change to: participant submitted code docker image.
+ Scoring program docker image:

(3) Participants’ docker
supplied at the time of code submission in the “Submit/View results” tab.

——
Desired behavior is:
1) If no organizer or participant docker: run all codes IPS in default docker.
2) If organizer supplies ingestion or scoring program docker, use those to run the corresponding I or S codes.
3) If organizer supplies participant code docker image, run the code of the participant in that docker instead of the default docker UNLESS the participants themselves supply a docker.

In conclusion, priorities are for code submitted:
- run in participant docker
- if no participant docker run in organizer supplied participant docker
- else run in default docker.

ABSOLUTE REPRODUCIBILITY: Once a submitted code ran in one given docker, we need to keep this info as metadata and ALWAYS run that submission in the same docker (whether this is a re-run or a code migration).
